File: workspaces/decision-platform/research/s1-4r-jax-risk/src/s1_4r_risk_research/_validation.py
"""NumPy/JAX public wrapper가 공유하는 host-side validation."""

from __future__ import annotations

import math
import re
from dataclasses import dataclass
from typing import NoReturn, TypeGuard, cast

import numpy as np
from numpy.typing import NDArray

from .errors import ResearchValidationError
from .models import EffectiveTrialProvenance, TransitionCounts

type FloatArray = NDArray[np.float64]
type IntArray = NDArray[np.int64]
type RealScalar = int | float | np.integer | np.floating

_REAL_SCALAR_TYPES = (int, float, np.integer, np.floating)
_INTEGER_SCALAR_TYPES = (int, np.integer)
_MAX_FLOAT64_INTEGER_MAGNITUDE = int(np.finfo(np.float64).max)
_SHA256_PATTERN = re.compile(r"^[0-9a-f]{64}$")
_PROVENANCE_METHODS = {
    "pre_registered_independent",
    "externally_estimated_effective_count",
}


@dataclass(frozen=True, slots=True)
class BacktestInputs:
    """검증이 끝난 loss/VaR와 strict exception sequence를 전달한다."""

    realized_losses: FloatArray
    forecast_vars: FloatArray
    exceptions: IntArray


def _raise(code: str) -> NoReturn:
    raise ResearchValidationError(code)


def _is_supported_real_scalar(value: object) -> TypeGuard[RealScalar]:
    return not isinstance(value, (bool, np.bool_)) and isinstance(value, _REAL_SCALAR_TYPES)


def validate_real_scalar(value: object, *, code: str = "research_input_invalid") -> float:
    """bool/complex를 제외한 finite real scalar를 Python float로 고정한다."""

    if not _is_supported_real_scalar(value):
        _raise(code)
    try:
        result = float(value)
    except (OverflowError, TypeError, ValueError):
        _raise(code)
    if not math.isfinite(result):
        _raise(code)
    return result


def validate_integer_scalar(value: object, *, code: str) -> int:
    """bool을 integer로 오인하지 않고 exact integer만 허용한다."""

    if isinstance(value, (bool, np.bool_)) or not isinstance(value, _INTEGER_SCALAR_TYPES):
        _raise(code)
    return int(value)


def validate_confidence(value: object) -> float:
    """Confidence는 finite real이며 open unit interval이어야 한다."""

    confidence = validate_real_scalar(value)
    if not 0.0 < confidence < 1.0:
        _raise("research_input_invalid")
    return confidence


def validate_significance(value: object) -> float:
    """Significance의 type, finite, domain 위반을 단일 stable code로 매핑한다."""

    significance = validate_real_scalar(value, code="significance_invalid")
    if not 0.0 < significance < 1.0:
        _raise("significance_invalid")
    return significance


def validate_sample_size(value: object) -> int:
    """PSR/DSR sample size의 type과 최소 길이 오류를 구분한다."""

    sample_size = validate_integer_scalar(value, code="research_input_invalid")
    if sample_size <= 1:
        _raise("research_input_too_short")
    if sample_size > _MAX_FLOAT64_INTEGER_MAGNITUDE:
        _raise("research_input_invalid")
    return sample_size


def validate_trial_count(value: object) -> int:
    """DSR trial count는 반올림하지 않은 integer N>=2여야 한다."""

    trial_count = validate_integer_scalar(value, code="trial_count_invalid")
    if trial_count < 2:
        _raise("trial_count_invalid")
    if trial_count > _MAX_FLOAT64_INTEGER_MAGNITUDE:
        _raise("trial_count_invalid")
    return trial_count


def _coerce_sequence(
    values: object,
    *,
    dimension_code: str = "research_input_invalid",
) -> FloatArray:
    """지원 container를 복사된 finite float64 1-D array로 변환한다."""

    if isinstance(values, np.ndarray):
        if values.ndim != 1:
            _raise(dimension_code)
        if values.dtype.kind not in "iuf":
            _raise("research_input_invalid")
        try:
            result = np.array(values, dtype=np.float64, copy=True)
        except (OverflowError, TypeError, ValueError):
            _raise("research_input_invalid")
    elif type(values) in (list, tuple):
        sequence = cast(list[object] | tuple[object, ...], values)
        try:
            shape_probe = np.asarray(sequence)
        except (OverflowError, TypeError, ValueError):
            _raise(dimension_code)
        if shape_probe.ndim != 1:
            _raise(dimension_code)
        if any(not _is_supported_real_scalar(value) for value in sequence):
            _raise("research_input_invalid")
        try:
            result = np.array(sequence, dtype=np.float64, copy=True)
        except (OverflowError, TypeError, ValueError):
            _raise("research_input_invalid")
        if result.ndim != 1:
            _raise(dimension_code)
    else:
        _raise("research_input_invalid")

    if not bool(np.all(np.isfinite(result))):
        _raise("research_input_invalid")
    return result


def validate_sequence(values: object, *, minimum_length: int = 1) -> FloatArray:
    """단일-sequence validation precedence를 적용한다."""

    result = _coerce_sequence(values)
    if result.size < minimum_length:
        _raise("research_input_too_short")
    return result


def validate_backtest_sequences(
    realized_losses: object,
    forecast_vars: object,
    *,
    minimum_length: int,
) -> BacktestInputs:
    """VaR 입력을 realized→forecast→shape→length 순서로 검증한다."""

    realized = _coerce_sequence(realized_losses)
    forecast = _coerce_sequence(forecast_vars, dimension_code="forecast_shape_invalid")
    if realized.shape != forecast.shape:
        _raise("forecast_shape_invalid")
    if realized.size < minimum_length:
        _raise("research_input_too_short")
    if bool(np.any(forecast < 0.0)):
        _raise("forecast_var_negative")
    exceptions = np.asarray(realized > forecast, dtype=np.int64)
    return BacktestInputs(realized, forecast, exceptions)


def validate_moment_pair(skewness: float, kurtosis: float) -> None:
    """Pearson moment inequality를 float64 roundoff tolerance와 함께 검사한다."""

    lower_bound = skewness * skewness + 1.0
    tolerance = (
        64.0
        * np.finfo(np.float64).eps
        * max(1.0, abs(kurtosis), abs(lower_bound))
    )
    if not math.isfinite(lower_bound) or kurtosis + tolerance < lower_bound:
        _raise("moment_invalid")


def validate_trial_provenance(
    provenance: object,
    *,
    trial_count: int,
) -> EffectiveTrialProvenance:
    """DSR effective-N provenance의 schema/count/frequency/digest를 원자 검증한다."""

    if not isinstance(provenance, EffectiveTrialProvenance):
        _raise("trial_provenance_invalid")
    if (
        not isinstance(provenance.schema_version, str)
        or provenance.schema_version != "s1.4r-effective-trials-v1"
    ):
        _raise("trial_provenance_invalid")
    if (
        not isinstance(provenance.method, str)
        or provenance.method not in _PROVENANCE_METHODS
    ):
        _raise("trial_provenance_invalid")
    try:
        raw_count = validate_integer_scalar(
            provenance.raw_trial_count,
            code="trial_provenance_invalid",
        )
        effective_count = validate_integer_scalar(
            provenance.effective_trial_count,
            code="trial_provenance_invalid",
        )
    except ResearchValidationError:
        _raise("trial_provenance_invalid")
    if not raw_count >= effective_count >= 2:
        _raise("trial_provenance_invalid")
    if effective_count != trial_count:
        _raise("trial_provenance_invalid")
    if not isinstance(provenance.sampling_frequency, str):
        _raise("trial_provenance_invalid")
    if not provenance.sampling_frequency.strip():
        _raise("trial_provenance_invalid")
    if not isinstance(provenance.trial_registry_sha256, str):
        _raise("trial_provenance_invalid")
    if _SHA256_PATTERN.fullmatch(provenance.trial_registry_sha256) is None:
        _raise("trial_provenance_invalid")
    variance_ddof = validate_integer_scalar(
        provenance.variance_ddof,
        code="trial_provenance_invalid",
    )
    if variance_ddof != 1:
        _raise("trial_provenance_invalid")
    return provenance


def transition_counts(exceptions: IntArray) -> TransitionCounts:
    """고정 길이 slice와 reduction으로 2-state transition counts를 계산한다."""

    previous = exceptions[:-1]
    current = exceptions[1:]
    return TransitionCounts(
        n00=int(np.sum((previous == 0) & (current == 0), dtype=np.int64)),
        n01=int(np.sum((previous == 0) & (current == 1), dtype=np.int64)),
        n10=int(np.sum((previous == 1) & (current == 0), dtype=np.int64)),
        n11=int(np.sum((previous == 1) & (current == 1), dtype=np.int64)),
    )


def validate_transition_identifiability(counts: TransitionCounts) -> None:
    """두 Markov row가 모두 관측되지 않으면 LRind를 산출하지 않는다."""

    if counts.n00 + counts.n01 == 0 or counts.n10 + counts.n11 == 0:
        _raise("insufficient_sample")

File: workspaces/decision-platform/research/s1-4x-numeric-parity/scala/src/main/scala/ai/trading/coach/s14x/core/Validation.scala
package ai.trading.coach.s14x.core

/**
 * production과 research API의 입력·오류 precedence를 한 경계에서 고정한다.
 * 성공 값은 immutable 입력과 positive-zero 결과만 다음 numeric 단계로 전달한다.
 */
object Validation:
  private val MaxProductionInputLength = 100000
  // IEEE-754 max finite의 exact integer 값이다. Decimal 문자열 반올림으로 상한을 만들지 않는다.
  private val MaxFloat64Integer = BigInt(2).pow(1024) - BigInt(2).pow(971)
  private val Sha256Pattern = "^[0-9a-f]{64}$".r
  private val ProvenanceMethods =
    Set("pre_registered_independent", "externally_estimated_effective_count")

  /** Typed core 경계에서 길이·finite precedence를 적용하고 immutable snapshot을 반환한다. */
  def productionSequence(
      values: Vector[Double],
      minimumLength: Int,
  ): Either[StableError, Vector[Double]] =
    if values.isEmpty then Left(StableError.InputEmpty)
    else if values.size > MaxProductionInputLength then Left(StableError.InputTooLong)
    else if values.size < minimumLength then Left(StableError.InputTooShort)
    else if values.exists(value => !value.isFinite) then Left(StableError.InputNonFinite)
    else Right(Vector.from(values))

  /** Transport-representable raw list의 shape → bool → type precedence를 보존한다. */
  def productionRawSequence(
      raw: Any,
      minimumLength: Int,
  ): Either[StableError, Vector[Double]] =
    raw match
      case _: Boolean => Left(StableError.InputBoolInvalid)
      case values: Vector[?] =>
        val nested = values.exists {
          case _: Iterable[?] => true
          case _: Array[?]    => true
          case _              => false
        }
        val boolean = values.exists {
          case _: Boolean => true
          case _          => false
        }
        val invalid = values.exists {
          case _: Byte | _: Short | _: Int | _: Long | _: BigInt | _: Double => false
          case _                                                             => true
        }
        if nested then Left(StableError.InputShapeInvalid)
        else if boolean then Left(StableError.InputBoolInvalid)
        else if invalid then Left(StableError.InputTypeInvalid)
        else
          val doubles = values.map {
            case value: Byte   => value.toDouble
            case value: Short  => value.toDouble
            case value: Int    => value.toDouble
            case value: Long   => value.toDouble
            case value: BigInt => value.toDouble
            case value: Double => value
            case _             => Double.NaN
          }
          productionSequence(doubles, minimumLength)
      case _ => Left(StableError.InputTypeInvalid)

  /** research 배열은 finite를 먼저 확인한 뒤 최소 길이를 확인해 frozen 오류 precedence를 보존한다. */
  def researchSequence(
      values: Vector[Double],
      minimumLength: Int = 1,
  ): Either[StableError, Vector[Double]] =
    if values.exists(value => !value.isFinite) then Left(StableError.ResearchInputInvalid)
    else if values.size < minimumLength then Left(StableError.ResearchInputTooShort)
    else Right(Vector.from(values))

  /** 연환산 period는 임의 정밀도 정수로 받아 양수 여부만 먼저 검증한다. */
  def positivePeriods(value: BigInt): Either[StableError, BigInt] =
    if value <= 0 then Left(StableError.PeriodsPerYearInvalid) else Right(value)

  /** production confidence를 열린 구간 (0, 1)로 제한하고 production stable code를 반환한다. */
  def productionConfidence(value: Double): Either[StableError, Double] =
    if !value.isFinite || value <= 0.0 || value >= 1.0 then Left(StableError.ConfidenceInvalid)
    else Right(value)

  /** research confidence를 열린 구간 (0, 1)로 제한하고 research 입력 오류로 통일한다. */
  def researchConfidence(value: Double): Either[StableError, Double] =
    if !value.isFinite || value <= 0.0 || value >= 1.0 then
      Left(StableError.ResearchInputInvalid)
    else Right(value)

  /** hypothesis-test significance를 finite 열린 구간 (0, 1)로 제한한다. */
  def significance(value: Double): Either[StableError, Double] =
    if !value.isFinite || value <= 0.0 || value >= 1.0 then
      Left(StableError.SignificanceInvalid)
    else Right(value)

  /** PSR 표본 수는 2 이상이면서 Float64로 유한하게 표현 가능한 정수만 허용한다. */
  def sampleSize(value: BigInt): Either[StableError, BigInt] =
    if value <= 1 then Left(StableError.ResearchInputTooShort)
    else if value > MaxFloat64Integer || !value.toDouble.isFinite then
      Left(StableError.ResearchInputInvalid)
    else Right(value)

  /** DSR 시행 수는 2 이상이면서 Float64 변환이 유한한 정수만 허용한다. */
  def trialCount(value: BigInt): Either[StableError, BigInt] =
    if value < 2 || value > MaxFloat64Integer || !value.toDouble.isFinite then
      Left(StableError.TrialCountInvalid)
    else Right(value)

  /** skewness/kurtosis의 Pearson 하한을 scale-aware ulp 허용오차와 함께 검증한다. */
  def momentPair(skewness: Double, kurtosis: Double): Either[StableError, Unit] =
    val lower = skewness * skewness + 1.0
    val tolerance =
      64.0 * Math.ulp(1.0) * math.max(1.0, math.max(math.abs(kurtosis), math.abs(lower)))
    if !skewness.isFinite || !kurtosis.isFinite || !lower.isFinite || kurtosis + tolerance < lower
    then Left(StableError.MomentInvalid)
    else Right(())

  /** DSR trial provenance의 schema·count·registry hash·ddof를 frozen 계약과 대조한다. */
  def provenance(value: TrialProvenance, trialCount: BigInt): Either[StableError, Unit] =
    val valid =
      value.schemaVersion == "s1.4r-effective-trials-v1" &&
        ProvenanceMethods.contains(value.method) &&
        value.rawTrialCount >= value.effectiveTrialCount &&
        value.effectiveTrialCount >= 2 &&
        value.effectiveTrialCount == trialCount &&
        value.samplingFrequency.trim.nonEmpty &&
        Sha256Pattern.matches(value.trialRegistrySha256) &&
        value.varianceDdof == 1
    if valid then Right(()) else Left(StableError.TrialProvenanceInvalid)

  /** production 결과의 finite 여부를 확인하고 성공한 zero의 부호를 정규화한다. */
  def finiteProduction(value: Double): Either[StableError, Double] =
    if value.isFinite then Right(NumericPrimitives.normalizeZero(value))
    else Left(StableError.ResultNonFinite)

  /** research 결과의 finite 여부를 확인하고 성공한 zero의 부호를 정규화한다. */
  def finiteResearch(value: Double): Either[StableError, Double] =
    if value.isFinite then Right(NumericPrimitives.normalizeZero(value))
    else Left(StableError.ResearchResultNonFinite)
